processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java

```diff
@@ -169,31 +191,35 @@ public int getNumRows()
       return positions.size();
     }
 
-    public Map<ByteBuffer, Integer> getPositions()
+    public Map<int[], Integer> getPositions()
     {
       return positions;
     }
 
-    private List<ByteBuffer> updateValues(
-        ByteBuffer key,
+    private List<int[]> updateValues(
+        int[] key,
         List<DimensionSelector> dims
     )
     {
-      if (dims.size() > 0) {
-        List<ByteBuffer> retVal = null;
-        List<ByteBuffer> unaggregatedBuffers = null;
+      int size = dims.size();
+      int index = key.length - size;
+      if (size > 0) {
+        List<int[]> retVal = null;
+        List<int[]> unaggregatedBuffers = null;
 
         final DimensionSelector dimSelector = dims.get(0);
         final IndexedInts row = dimSelector.getRow();
         if (row == null || row.size() == 0) {
-          ByteBuffer newKey = key.duplicate();
-          newKey.putInt(dimSelector.getValueCardinality());
-          unaggregatedBuffers = updateValues(newKey, dims.subList(1, dims.size()));
+          key[index] = dimSelector.getValueCardinality();
+          unaggregatedBuffers = updateValues(key, dims.subList(1, size));
+        } else if (row.size() == 1) {
+          key[index] = row.get(0);
+          unaggregatedBuffers = updateValues(key, dims.subList(1, size));
         } else {
           for (Integer dimValue : row) {
-            ByteBuffer newKey = key.duplicate();
-            newKey.putInt(dimValue);
-            unaggregatedBuffers = updateValues(newKey, dims.subList(1, dims.size()));
+            int[] newKey = Arrays.copyOf(key, key.length);
```

Comment:
I think in the most of cases, `key.duplicate()` (1 ref, 5 int, 1 long, 1 boolean) uses more memory than simple `int[]`.